### ISSUE TITLE
[Help][Finetuning] Single-GPU finetuning example does not save the model but completes successfully

### ISSUE THREAD

**Issue by animeshj9** [meta-llama/llama-recipes#668]
### System Info

CUDA 12.6, 1 H100 80GB 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

1) Followed the quickstart here for [singlegpu_finetuning.md](https://github.com/meta-llama/llama-recipes/blob/main/recipes/quickstart/finetuning/singlegpu_finetuning.md). 

I am able to run the default alpaca-dataset using the command:

`FSDP_CPU_RAM_EFFICIENT_LOADING=1 \
python finetuning.py --quantization 4bit \
    --model_name pretrained_model \
    --tokenizer_name tokenizer \
    --use_peft --peft_method lora \
    --output_dir funetuned_model \
    --dataset alpaca_dataset \
    --save_model`

The dataset file is stored in src/llama_recipes/datasets folder and the model gets saved fine. 
_Model and Tokenizer downloaded from hugging face locally for "meta-llama/Meta-Llama-3.1-8B-Instruct" using the script [here](https://gist.github.com/animeshj9/5cea5ca63cf86356d17582c048dcd1c6)_

2) Now I changed the dataset with my own dataset in the same { "instruction": "", "input": "", "output": "" } alpaca format. Upon running the finetuning script with the same command again, the script completes but does *not* save the model anywhere.

I suspect it might be related to the avg_eval_loss showing as inf:
![image](https://github.com/user-attachments/assets/730334c5-0ca2-48ba-a68b-66653680e7a7)

_Notes about the dataset:
- the input is always a chunk of C/C++ code also some XML, json, text as well.
- the output contains a summary, or other metadata about that code_

### Error logs

No error message. The model just does not save.

avg_train_prep = nan
avg_train_loss = nan
avg_eval_prep = nan
avg_eval_loss = inf

### Expected behavior

The finetune completes, giving out a eval loss and saving the model (just as what happens when I run with the alpaca dataset at https://raw.githubusercontent.com/tatsu-lab/stanford_alpaca/main/alpaca_data.json)

**Comment by init27** [comment]
cc: @wukaixingxp

**Comment by mreso** [comment]
Hi @animeshj9 the training script compares the eval loss after each epoch and [only saves the model if its smaller](url). Initial loss is Inf and comparing it against NaN will always be False.
My assumption is that even if you save the model by removing this check it will not be of much value something went wrong during training as your loss is NaN. You'll need to debug your hyper parameters and input data on why your loss becomes NaN. Maybe start with triggering a breakpoint in your loop if your loss gets NaN for the first time and them look at the data fed to the model.

Something to the effect of:
```
if torch.isnan(loss).any():
    breakpoint()
```

**Comment by animeshj9** [comment]
Thanks @mreso . I will try and update here.

**Comment by animeshj9** [comment]
I was able to run this on a different dataset with fewer tokens in the "input" and it seems to have worked.  Guess the issue is with our data. Thank you for your help. Closing this for now.
